File: App/Java/app/src/main/java/com/example/tonjies/templateproject/module/weather/WeatherActivity.java
package com.example.tonjies.templateproject.module.weather;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.example.common.util.ACatch;
import com.example.common.util.L;
import com.example.tonjies.templateproject.R;
import com.example.tonjies.templateproject.module.select.SelectActivity;
import com.example.tonjies.templateproject.module.weather.bean.Weather;
import com.example.tonjies.templateproject.module.weather.presenter.WeatherPresenter;
import com.example.tonjies.templateproject.module.weather.presenter.contract.WeatherContract;
import com.scwang.smartrefresh.layout.SmartRefreshLayout;
import com.scwang.smartrefresh.layout.api.RefreshLayout;
import com.scwang.smartrefresh.layout.listener.OnRefreshListener;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

/**
 * Created by 舍长
 * describe: 天气显示界面
 */
public class WeatherActivity extends AppCompatActivity implements WeatherContract.WeatherView {

    //温度
    @BindView(R.id.txtTemp)
    TextView txtTemp;
    //城市
    @BindView(R.id.txtCity)
    TextView txtCity;
    //天气信息
    @BindView(R.id.txtInfo)
    TextView txtInfo;
    //
    @BindView(R.id.forecastLayout)
    LinearLayout forecastLayout;
    @BindView(R.id.forecast)
    LinearLayout forecast;

    //AQI指数
    @BindView(R.id.txtAqi)
    TextView txtAqi;

    //Pm2.5指数
    @BindView(R.id.txtPm25)
    TextView txtPm25;

    //舒适度
    @BindView(R.id.txtComfort)
    TextView txtComfort;

    //洗车指数
    @BindView(R.id.txtWash)
    TextView txtWash;

    //运动指数
    @BindView(R.id.txtSport)
    TextView txtSport;


    //城市选择页面
    @BindView(R.id.imgPlace)
    ImageView imgPlace;

    //下拉刷新
    @BindView(R.id.refreshLayout)
    SmartRefreshLayout refreshLayout;

    //P层接口
    private WeatherPresenter weatherIPresenter;

    //天气id
    private String weatherId;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_weather);
        ButterKnife.bind(this);
        init();
    }

    private void init() {
        weatherIPresenter = new WeatherPresenter(this);
        //查看是否已经存储了天气的数据
        weatherId = ACatch.getString("weatherId", "");
        //不等于空，说明之前已经请求过了，直接拿着之前请求过的天气id去气球天气数据即可
        if (!TextUtils.isEmpty(weatherId)) {
            weatherIPresenter.requestWeather(weatherId);
            L.d("从存储库中取出的:" + weatherId);
        }
        weatherIPresenter.requestWeather(weatherId);
        refreshLayout.setVisibility(View.GONE);
        //头部刷新
        refreshLayout.setOnRefreshListener(new OnRefreshListener() {
            @Override
            public void onRefresh(RefreshLayout refreshlayout) {
                refreshLayout.getLayout().postDelayed(new Runnable() {
                    @Override
                    public void run() {
                        weatherIPresenter.requestWeather(weatherId);
                        refreshLayout.finishRefresh();
                        Toast.makeText(WeatherActivity.this, "更新完毕", Toast.LENGTH_SHORT).show();
                    }
                }, 400);
            }
        });
    }

    @SuppressLint("SetTextI18n")
    @Override
    public void getData(Weather weather) {
        refreshLayout.setVisibility(View.VISIBLE);
        Weather.HeWeatherBean heWeatherBean = weather.getHeWeather().get(0);
        //城市名
        txtCity.setText(heWeatherBean.getBasic().getCity());
        //温度
        txtTemp.setText(heWeatherBean.getNow().getTmp() + "℃");
        //天气信息
        txtInfo.setText(heWeatherBean.getNow().getCond_txt());
        //aql指数
        txtAqi.setText(heWeatherBean.getAqi().getCity().getAqi());
        //pm2.5
        txtPm25.setText(heWeatherBean.getAqi().getCity().getPm25());
        //舒适度
        txtComfort.setText("舒适度：" + heWeatherBean.getSuggestion().getComf().getTxt());
        //洗车指数
        txtWash.setText("洗车指数:" + heWeatherBean.getSuggestion().getCw().getTxt());
        //运动建议
        txtSport.setText("运动建议:" + heWeatherBean.getSuggestion().getSport().getTxt());
        //最近几天的天气状况
        List<Weather.HeWeatherBean.DailyForecastBean> daily_forecast = heWeatherBean.getDaily_forecast();
        forecastLayout.removeAllViews();
        for (int i = 0; i < daily_forecast.size(); i++) {
            Weather.HeWeatherBean.DailyForecastBean dailyForecastBean = daily_forecast.get(i);
            String data = dailyForecastBean.getDate();//日期
            String cond = dailyForecastBean.getCond().getTxt_d();//天气状况
            String max = dailyForecastBean.getTmp().getMax();//最高温度
            String min = dailyForecastBean.getTmp().getMin();//最低温度
            View view = LayoutInflater.from(this).inflate(R.layout.forecast_item, forecastLayout, false);
            TextView tvDate = view.findViewById(R.id.txtDate);
            TextView tvInfo = view.findViewById(R.id.txtInfo);
            TextView maxText = view.findViewById(R.id.txtMax);
            TextView minText = view.findViewById(R.id.txtMin);

            tvDate.setText(data);
            tvInfo.setText(cond);
            maxText.setText(max);
            minText.setText(min);
            forecastLayout.addView(view);
        }
    }

    //跳转到城市选择页面
    @OnClick(R.id.imgPlace)
    public void onViewClicked() {
        startActivity(new Intent(this, SelectActivity.class));
    }
}

File: App/kotlins/app/src/main/java/com/example/tonjies/weatherx/module/weather/WeatherActivity.kt
package com.example.tonjies.weatherx.module.weather

import android.graphics.Color
import android.os.Build
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.widget.TextView
import com.example.tonjies.weatherx.R
import com.example.tonjies.weatherx.R.id.*
import com.example.tonjies.weatherx.base.MvpActivity
import com.example.tonjies.weatherx.module.select.SelectActivity
import com.example.tonjies.weatherx.module.weather.bean.Weather
import com.example.tonjies.weatherx.module.weather.presenter.WeatherPresenter
import com.example.tonjies.weatherx.module.weather.view.WeatherView
import com.example.tonjies.weatherx.util.AppManger
import com.example.tonjies.weatherx.util.L
import com.example.tonjies.weatherx.util.ShareUtils
import kotlinx.android.synthetic.main.activity_weather.*
import kotlinx.android.synthetic.main.item_city.*
import org.jetbrains.anko.find
import org.jetbrains.anko.startActivity
import org.jetbrains.anko.toast
import org.w3c.dom.Text

/**
 * 天气详情页面
 */
class WeatherActivity : MvpActivity<WeatherPresenter>(), WeatherView {

    //P层
    private var weatherPresenter = WeatherPresenter()

    private var pressTime: Long = 0

    //WeatheId
    private var weatherId: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        //将背景延伸到状态栏
        if (Build.VERSION.SDK_INT >= 21) {
            val decorView = window.decorView
            decorView.systemUiVisibility = View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN or View.SYSTEM_UI_FLAG_LAYOUT_STABLE
            window.statusBarColor = Color.TRANSPARENT
        }
        setContentView(R.layout.activity_weather)
        weatherId = ShareUtils.getString(this, "weatherId", "")
        if (weatherId != "") {
            weatherPresenter.requestWeather(weatherId!!)
        } else {
            val intent = intent
            weatherId = intent.getStringExtra("weatherId")
        }
        weatherPresenter.mView = this
        imgPlace.setOnClickListener {
            startActivity<SelectActivity>()
        }
        swipeRefresh.setOnRefreshListener {
            toast("当前已经是最新数据")
            weatherPresenter.requestWeather(weatherId!!)
        }
    }

    /**
     * 设置天气数据
     */
    override fun setData(weather: Weather) {
        swipeRefresh.visibility = View.VISIBLE//显示页面布局
        swipeRefresh.isRefreshing = false
        val cityName = getWeather(weather).basic.city//城市名
//        val updateTime: String = getWeather(weather).update.loc.split(" ")[1]//更新时间
        val degree = getWeather(weather).now.tmp + "℃"//温度
        val weatherInfo = getWeather(weather).now.cond_txt//天气

        txtCity.text = cityName//城市
        txtTemp.text = degree//当前温度
        txtInfo.text = weatherInfo//当前天气状况

        val aqi = getWeather(weather).aqi.city.aqi //aql
        val pm25 = getWeather(weather).aqi.city.pm25//pm2.5
        L.d("sss$aqi,$pm25")
        txtAqi.text = aqi//aqi指数
        txtPm25.text = pm25//pm25

        val comfort = "舒适度：${getWeather(weather).suggestion.comf.txt}"
        val carWash = "洗车指数:${getWeather(weather).suggestion.cw.txt}"
        val sport = "运动建议：${getWeather(weather).suggestion.sport.txt}"

        txtComfort.text = comfort
        txtWash.text = carWash
        txtSport.text = sport

        //最近几天的天气状况
        val daily_forecast = getWeather(weather).daily_forecast
        forecastLayout.removeAllViews()
        for (i in 0..daily_forecast.size) {
            val data = daily_forecast[i].date//日期
            val cond = daily_forecast[i].cond.txt_d//天气状况
            val max = daily_forecast[i].tmp.max//最高温度
            val min = daily_forecast[i].tmp.min//最低温度
//            L.d("日期是：$data 天气状况:$cond 最高温度:$max 最低温度:$min")
            val view = LayoutInflater.from(this).inflate(R.layout.forecast_item, forecastLayout, false)
            val dateText: TextView = view.find(R.id.txtDate)//时间
            val infoText: TextView = view.find(R.id.txtInfo)
            val maxText: TextView = view.find(R.id.txtMax)
            val minText: TextView = view.find(R.id.txtMin)

            dateText.text = data
            infoText.text = cond
            maxText.text = max
            minText.text = min
            forecastLayout.addView(view)
        }


    }


    fun getWeather(weather: Weather) =
            weather.HeWeather[0]


    override fun onBackPressed() {
        val time = System.currentTimeMillis()//当前时间
        if (time - pressTime > 2000) {
            toast("再按一次退出程序")
            pressTime = time
        } else {
            AppManger.exitApp(this)
        }
    }
}
